Check STID length is valid

spragchris opened the pull request:
Check that length of STID is long enough to avoid exceptions while in menus

spragchris committed:
Check STID length is valid

Check that length of STID is long enough to avoid exceptions while in menus

spragchris commented:
Hi, there is an exception being thrown every time the loop checked RS data, while in menus. It was being caught higher up, but could be avoided by checking the length before returning a valid status for the data. 

This exception can be seen in the output window, or diagnostic tools window.

kokolihapihvi commented:
Hey, thanks for the PR!

You missed the other variation of the method, and have you verified that no valid STID length is less than 16? I don't remember exactly how long the structs usually are.

spragchris commented:
@kokolihapihvi I see what you mean, I think perhaps a better solution would be to avoid building 'Learn a Song' data when in menus (a higher level change). I'll modify the PR tonight.

spragchris committed:
Move song list check to higher level

When no song is present, the STID is too short to retrieve sound bank details, so fail early with return instead of allowing exception on array access.

spragchris committed:
Revert previous changes

Reverting of previous changes in branch

spragchris commented:
@kokolihapihvi I have reverted previous changes, and relocated the check to ReadHIRC, where an early return prevents the exception when accessing stid[16].

kokolihapihvi commented:
Hi, sorry for the confusion, let me try to explain better..

What I meant by "you missed the other variation of the method" is that you added the check to `IsValidHIRCPointer(IntPtr ptr, out int hLen, out int sLen)`, but not `IsValidHIRCPointer(IntPtr ptr)`.

You could probably get rid of some duplicate code too by calling `IsValidHIRCPointer(IntPtr ptr, out int hLen, out int sLen)` from inside `IsValidHIRCPointer(IntPtr ptr)` and just discarding the out parameters. (I'm not sure why I didn't do that in the first place)

I think the length check you are trying to add should be in the IsValidHIRCPointer functions instead of higher up. Like you had it in the first commit.

And my question was; how did you get to 16 and does it work as intended with the check in place?

Also to clarify what the pointer actually is: when a songs preview audio is playing, it's loaded into memory in a HIRC struct, and inside that struct there is a STID struct which contains the file name, which is "Song_something", where the "something" part is the song id. This struct only exists in memory when a song preview is playing, which would probably explain why its throwing exceptions.

The point of the ReadHIRC function is to find that struct in memory, and find the STID struct after it, and then read the file name out of it to determine the song id of the currently selected song.

Hope this helps!

spragchris commented:
I don't know, its actually a tough one (especially with my only superficial knowledge). An exception gets thrown in the condition that the stid value is 4 bytes long, when attempting to access index 16 (so I suppose checking for 17 length would have been correct in my suggestion.)

Perhaps checking for 4 bytes would be best, as even a 17 byte long array will still be accessed at byte 18 on the next statement when attempting to find the song.

I moved the check into this file because it's closer to the point of the error, and doesn't assume anything about the purposes or usages of the other two methods.

I'll check in a fix to look for 4 bytes long to indicate the wrong size STID to find a song within.

I can make any other change you would suggest, again I must admit limited knowledge. I simply noticed an exception was being thrown (and caught and discarded) regularly in the output window.

spragchris committed:
More specific guard condition

As a STID of 4 bytes long will not contain song data, quit out of song check before receiving an exception attempting to access song data.
